Source/WebCore/rendering/GridMasonryLayout.cpp

```diff
@@ -0,0 +1,167 @@
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "config.h"
+#include "GridMasonryLayout.h"
+
+#include "GridLayoutFunctions.h"
+#include "GridPositionsResolver.h"
+#include "WritingMode.h"
+
+namespace WebCore {
+
+void GridMasonryLayout::performMasonryPlacement(const HashMap<RenderBox*, GridArea>& firstTrackItems, const Vector<RenderBox*> itemsWithDefiniteGridAxisPosition, const Vector<RenderBox*> itemsWithIndefinitePosition, GridTrackSizingDirection masonryAxisDirection)
+{
+    m_masonryAxisDirection = masonryAxisDirection;
+    m_gridAxisDirection = GridPositionsResolver::gridAxisDirection(m_masonryAxisDirection);
+    m_masonryAxisGridGap = m_renderGrid.gridGap(m_masonryAxisDirection);
+    m_gridAxisTracksCount = m_renderGrid.numTracks(m_gridAxisDirection);
+    resizeAndResetRunningPositions(); 
+
+
+    // 2.4 Masonry Layout Algorithm
+    addItemsToFirstTrack(firstTrackItems);
+    placeItemsWithDefiniteGridAxisPosition(itemsWithDefiniteGridAxisPosition);
+    placeItemsWithIndefiniteGridAxisPosition(itemsWithIndefinitePosition);
+}
+
+void GridMasonryLayout::resizeAndResetRunningPositions()
+{
+    m_runningPositions.resize(m_gridAxisTracksCount);
+    m_runningPositions.fill(LayoutUnit());
+}
+
+void GridMasonryLayout::addItemsToFirstTrack(const HashMap<RenderBox*, GridArea>& firstTrackItems)
+{
+    for (auto& [item, gridArea] : firstTrackItems)
+        insertIntoGridAndLayoutItem(item, gridArea);
+}
+
+void GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithDefiniteGridAxisPosition)
+{
+    for (auto* item : itemsWithDefiniteGridAxisPosition) {
+        auto itemSpan = m_renderGrid.currentGrid().gridItemSpan(*item, m_gridAxisDirection);
+
+        ASSERT(!itemSpan.isIndefinite());
+
+        itemSpan.translate(m_renderGrid.currentGrid().explicitGridStart(m_gridAxisDirection));
+
+        GridArea area = m_masonryAxisDirection == ForRows ? GridArea(m_masonryAxisSpan, itemSpan) : GridArea(itemSpan, m_masonryAxisSpan);
+        insertIntoGridAndLayoutItem(item, area);
+    }
+}
+
+void GridMasonryLayout::placeItemsWithIndefiniteGridAxisPosition(const Vector<RenderBox*>& itemsWithIndefinitePosition)
+{
+    for (auto* item : itemsWithIndefinitePosition)
+        insertIntoGridAndLayoutItem(item, nextMasonryPositionForItem(item));
+}
+
+void GridMasonryLayout::setItemGridAxisContainingBlockToGridArea(RenderBox* child)
+{
+    if (m_gridAxisDirection == ForColumns)
+        child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForColumns));
+    else
+        child->setOverridingContainingBlockContentLogicalHeight(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, ForRows));
```

Comment:
Still curious why this isn't just:

`
child->setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(*child, m_gridAxisDirection));
`

Reply:
That works if the grid axis is in the column direction because we want to update the height. If it's in the row direction we need to update the height instead. (Or are you talking more about replacing the `ForColumns`/`ForRows` with `m_gridAxisDirection`)